backend/api/EventHandlers/MqttEventHandler.cs

```diff
@@ -55,33 +60,161 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+            var robot = await RobotService.ReadByName(isarRobotInfo.RobotName);
+
+            if (robot == null)
+            {
                 _logger.LogInformation(
-                    "ISAR instance for robot {name} with id {id} is connected. Robot is enabled and host ({host}) and port ({port}) is updated.",
-                    isarRobot.RobotId,
-                    robot.Id,
-                    isarRobot.Host,
-                    isarRobot.Port
+                    "Received message from new ISAR instance with robot name {name}. Adding new robot to database.",
+                    isarRobotInfo.RobotName
                 );
+
+                var robotQuery = new CreateRobotQuery()
+                {
+                    Name = isarRobotInfo.RobotName,
+                    Model = isarRobotInfo.RobotModel,
+                    SerialNumber = isarRobotInfo.SerialNumber,
+                    VideoStreams = isarRobotInfo.VideoStreamQueries,
+                    Host = isarRobotInfo.Host,
+                    Port = isarRobotInfo.Port,
+                    Status = RobotStatus.Available,
+                    Enabled = true
+                };
+
+                robot = await RobotService.Create(robotQuery);
+                _logger.LogInformation("Added robot {name} to database", robot.Name);
             }
+            else if (RobotHasSignificantChange(robot, isarRobotInfo))
+            {
+                _logger.LogInformation(
+                    "A change was discovered on robot {name} and the database will be updated",
+                    robot.Name
+                );
+
+                var existingVideoStreams = new List<string>();
+                foreach (var existingVideoStream in robot.VideoStreams)
+                    existingVideoStreams.Add(existingVideoStream.Name);
+
+                foreach (var videoStreamQuery in isarRobotInfo.VideoStreamQueries)
+                {
+                    if (!existingVideoStreams.Contains(videoStreamQuery.Name))
+                    {
+                        var videoStream = new VideoStream
+                        {
+                            Name = videoStreamQuery.Name,
+                            Url = videoStreamQuery.Url,
+                            Type = videoStreamQuery.Type
+                        };
+                        robot.VideoStreams.Add(videoStream);
+                    }
+                }
+
+                robot.Host = isarRobotInfo.Host;
+                robot.Port = isarRobotInfo.Port;
+
+                robot = await RobotService.Update(robot);
+                _logger.LogInformation("Updated robot {name} in database", robot.Name);
+            }
+        }
+
+        private bool RobotHasSignificantChange(Robot robot, IsarRobotInfoMessage isarRobotInfo)
+        {
+            if (robot.Host != isarRobotInfo.Host || robot.Port != isarRobotInfo.Port)
+            {
+                _logger.LogInformation("Halleluja");
```

Comment:
Quite some repeating code here, is it possible to make this one method, both updating if needed and alerting us of it in the same method?

Reply:
I have been looking quite a bit at this previously and it is the checking of the video stream object that has me confused. I agree of course but haven't found a neater way yet.

Reply:
I am leaving this as it is even though I know it's not the prettiest thing I have produced... Suggestions for how to change it would be welcome.